### ISSUE TITLE
Crash when creating zcm object in Python 3

### ISSUE THREAD

**Issue by volalto** [ZeroCM/zcm#183]
I have been trying to make zcm work with python3 (cython3) on two different systems (Debian 9 - Python 3.5 and Arch - Python 3.6).

I was able to compile the library but I get a core dump when I create a zcm object:
```python
import sys
sys.path.insert(0, '/opt/zcm/lib/python3.6/site-packages')
from zcm import ZCM
zcm = ZCM("ipc")
```
On the Debian machine zcm for Python 2 (compiled with the same flags as Python 3) works as expected.

Core dump:
```               
           PID: 21404 (python)
           UID: 1000 (volalto)
           GID: 1000 (volalto)
        Signal: 11 (SEGV)
     Timestamp: Tue 2018-04-17 22:57:57 CEST (9min ago)
  Command Line: python sub_test.py
    Executable: /usr/bin/python3.6
 Control Group: /user.slice/user-1000.slice/session-c2.scope
           PID: 21404 (python)
           UID: 1000 (volalto)
           GID: 1000 (volalto)
        Signal: 11 (SEGV)
     Timestamp: Tue 2018-04-17 22:57:57 CEST (9min ago)
  Command Line: python sub_test.py
    Executable: /usr/bin/python3.6
 Control Group: /user.slice/user-1000.slice/session-c2.scope
          Unit: session-c2.scope
         Slice: user-1000.slice
       Session: c2
     Owner UID: 1000 (volalto)
       Boot ID: 1fd14f7215ff4042baa7894facb27283
    Machine ID: f6995e53258d49ac989791c2c5a9bb96
      Hostname: voldesktop
       Storage: /var/lib/systemd/coredump/core.python.1000.1fd14f7215ff4042baa7894facb27283.21404.1523998677000000.lz4
       Message: Process 21404 (python) of user 1000 dumped core.
                
                Stack trace of thread 21404:
                #0  0x00007ff063c306a0 zcm_try_stop (libzcm.so)
                #1  0x00007ff063e53c62 __pyx_pw_3zcm_3ZCM_25stop (zcm.cpython-36m-x86_64-linux-gnu.so)
                #2  0x00007ff063e50225 __Pyx_PyObject_CallNoArg (zcm.cpython-36m-x86_64-linux-gnu.so)
                #3  0x00007ff063e580a8 __pyx_tp_dealloc_3zcm_ZCM (zcm.cpython-36m-x86_64-linux-gnu.so)
                #4  0x00007ff063e5391a __pyx_tp_new_3zcm_ZCM (zcm.cpython-36m-x86_64-linux-gnu.so)
                #5  0x00007ff064e26235 n/a (libpython3.6m.so.1.0)
                #6  0x00007ff064dd9ee5 _PyObject_FastCallDict (libpython3.6m.so.1.0)
                #7  0x00007ff064df6e4e n/a (libpython3.6m.so.1.0)
                #8  0x00007ff064d8ab5a _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
                #9  0x00007ff064df72c8 PyEval_EvalCodeEx (libpython3.6m.so.1.0)
                #10 0x00007ff064d8a6ec PyEval_EvalCode (libpython3.6m.so.1.0)
                #11 0x00007ff064e5cf84 n/a (libpython3.6m.so.1.0)
                #12 0x00007ff064e5f751 PyRun_FileExFlags (libpython3.6m.so.1.0)
                #13 0x00007ff064e5f954 PyRun_SimpleFileExFlags (libpython3.6m.so.1.0)
                #14 0x00007ff064e5570b Py_Main (libpython3.6m.so.1.0)
                #15 0x000055fb7c648c39 main (python3.6)
                #16 0x00007ff0651eff4a __libc_start_main (libc.so.6)
                #17 0x000055fb7c648dba _start (python3.6)
```

**Comment by jbendes** [comment]
Are you sure you ran `cython3` instead of `cython`? That tool compiles a native library that is used by python. I wouldn't expect `cython` to work with `python3`

**Comment by volalto** [comment]
Hi @jbendes. I'm indeed quite sure.

In Debian I changed all occurrences of "cython" to "cython3" in waf scripts and also  tried running the compilation in a python3 virtualenv.

To make sure I also tried on my Arch box where python3 is the only python available and "cython" is just a link to cython3.

Both methods resulted in a library being generated in the right path and imported by python3 scripts with no complaints (after defining LD_LIBRARY_PATH to the location of zcm.so).

**Comment by jbendes** [comment]
Was able to replicate your issue. Was a small bug in the zcm.pyx cython file. Zcm assumed construction never failed, which is an invalid assumption. Should be fixed. There are of course some more changes between python 2 and 3, but the fix I pushed up on `core-python3` should be enough to get you going. Let me know what you think. Also would be much appreciated if you wanted to push up a branch once you get everything working with some edits to zcm.pyx and any of the examples to get this python3 to be a fully supported language!

**Comment by volalto** [comment]
Thanks @jbendes! I already gave it a try and I was able to run my simple example! 

There is still an issues with passing python3 strings to cython. The library complains about expecting bytes and getting str. This can be temporary fixed by declaring all string (e.g. url) as b'string'.

I'm a bit confused as I understood from cython documentation that str should be fine with both Python2 and 3. I also tried to force Python3 syntax (cython -3) but I still get the error.

I'll definitely contribute back any improvement on the Python port.
